[CI] `linux://python/ray/serve/tests:test_cli` is failing/flaky on master.

Noticed the are some flakiness in `test_cli` across the last 100 commits. The issue is consistently related to `python/ray/serve/tests/test_cli.py::test_deploy_gang_scaling[4-2]` with the trace below: 


```

[2026-04-16T19:09:57Z] ________________________ test_deploy_gang_scaling[4-2] _________________________
--
[2026-04-16T19:09:57Z]
[2026-04-16T19:09:57Z] serve_instance = <ray.serve._private.client.ServeControllerClient object at 0x7fdbbf0bbe30>
[2026-04-16T19:09:57Z] initial_replicas = 4, final_replicas = 2
[2026-04-16T19:09:57Z]
[2026-04-16T19:09:57Z]     @pytest.mark.parametrize(
[2026-04-16T19:09:57Z]         "initial_replicas, final_replicas",
[2026-04-16T19:09:57Z]         [
[2026-04-16T19:09:57Z]             (4, 2),  # Downscale: 2 gangs -> 1 gang
[2026-04-16T19:09:57Z]             (2, 4),  # Upscale:   1 gang  -> 2 gangs
[2026-04-16T19:09:57Z]         ],
[2026-04-16T19:09:57Z]     )
[2026-04-16T19:09:57Z]     def test_deploy_gang_scaling(serve_instance, initial_replicas, final_replicas):
[2026-04-16T19:09:57Z]         """Test gang-aware scaling via serve deploy preserves existing gangs."""
[2026-04-16T19:09:57Z]         gang_size = 2
[2026-04-16T19:09:57Z]         config_files = {
[2026-04-16T19:09:57Z]             replicas: os.path.join(
[2026-04-16T19:09:57Z]                 os.path.dirname(__file__),
[2026-04-16T19:09:57Z]                 "test_config_files",
[2026-04-16T19:09:57Z]                 f"gang_scheduling_{replicas}_replicas.yaml",
[2026-04-16T19:09:57Z]             )
[2026-04-16T19:09:57Z]             for replicas in (initial_replicas, final_replicas)
[2026-04-16T19:09:57Z]         }
[2026-04-16T19:09:57Z]
[2026-04-16T19:09:57Z]         # First deploy.
[2026-04-16T19:09:57Z]         subprocess.check_output(["serve", "deploy", config_files[initial_replicas]])
[2026-04-16T19:09:57Z]         app_url = get_application_url(app_name="gang_app")
[2026-04-16T19:09:57Z]         wait_for_condition(lambda: httpx.post(f"{app_url}/").status_code == 200, timeout=30)
[2026-04-16T19:09:57Z]
[2026-04-16T19:09:57Z]         # Collect initial gang IDs.
[2026-04-16T19:09:57Z]         initial_num_gangs = initial_replicas // gang_size
[2026-04-16T19:09:57Z]         initial_gang_ids = set()
[2026-04-16T19:09:57Z]         for _ in range(initial_replicas * 10):
[2026-04-16T19:09:57Z]             data = json.loads(httpx.post(f"{app_url}/").text)
[2026-04-16T19:09:57Z]             if data["gang_id"] is not None:
[2026-04-16T19:09:57Z]                 initial_gang_ids.add(data["gang_id"])
[2026-04-16T19:09:57Z] >       assert len(initial_gang_ids) == initial_num_gangs
[2026-04-16T19:09:57Z] E       AssertionError: assert 1 == 2
[2026-04-16T19:09:57Z] E        +  where 1 = len({'529feb30'})
[2026-04-16T19:09:57Z]
[2026-04-16T19:09:57Z] python/ray/serve/tests/test_cli.py:860: AssertionError
 

<br class="Apple-interchange-newline">
```
 
- 64385a50c7e3da838c64ffd96cdacf58a33ddf62 FLAKY [Buildkite :ray-serve: serve: direct ingress tests (same loop) [g16_s17]](https://buildkite.com/ray-project/postmerge/builds/17121#019d9efc-7ef2-4138-96ac-6551728f51c1)
- e14b30fd2e0550e31cf4a9083bf584375bb6f460 FLAKY [Buildkite :ray-serve: serve: same event loop tests [g16_s5]](https://buildkite.com/ray-project/postmerge/builds/17096#019d99d3-f963-4acf-ba2d-352c4296f0c9)
- 131d39c79ca5793f54de15d6d7b1b69a8963d0b2 FLAKY [Buildkite :ray-serve: serve: same event loop tests [g16_s5]](https://buildkite.com/ray-project/postmerge/builds/17076#019d97e8-0a78-4651-a482-5761b10daf3b)
- 56f7ec1c4785abd5168bb0329f7284d308630fa3 FLAKY [Buildkite :ray-serve: serve: python 3.12 tests (0) [g16_s6]](https://buildkite.com/ray-project/postmerge/builds/17069#019d9734-5693-43e6-90cc-272661ba64c4)
- 56f7ec1c4785abd5168bb0329f7284d308630fa3 FLAKY [Buildkite :ray-serve: serve: gRPC tests [g16_s4]](https://buildkite.com/ray-project/postmerge/builds/17069#019d9734-5691-403c-b0d0-4cc70b5d3afb)
- 56f7ec1c4785abd5168bb0329f7284d308630fa3 FLAKY [Buildkite :ray-serve: serve: same event loop tests [g16_s5]](https://buildkite.com/ray-project/postmerge/builds/17069#019d9734-5692-4809-8db0-4dcd1f85b97f)
- f46a950257f9244a9181675b9b3f9d430c06c795 FLAKY [Buildkite :ray-serve: serve: direct ingress tests (same loop) [g16_s17]](https://buildkite.com/ray-project/postmerge/builds/17057#019d939e-bbdc-4d99-946b-7d7a82baae41)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-linux://python/ray/serve/tests:test_cli-END
....

--- a/python/ray/serve/tests/test_cli.py
+++ b/python/ray/serve/tests/test_cli.py
@@ -11,9 +11,10 @@ import httpx
 import pytest
 import yaml
 
+import ray
 from ray import serve
 from ray._common.test_utils import wait_for_condition
-from ray.serve._private.common import DeploymentID
+from ray.serve._private.common import DeploymentID, ReplicaState
 from ray.serve._private.constants import SERVE_DEFAULT_APP_NAME
 from ray.serve._private.test_utils import get_application_url
 from ray.serve.scripts import remove_ansi_escape_sequences
@@ -836,6 +837,8 @@ def test_deploy_gang_scheduling(serve_instance):
 def test_deploy_gang_scaling(serve_instance, initial_replicas, final_replicas):
     """Test gang-aware scaling via serve deploy preserves existing gangs."""
     gang_size = 2
+    deployment_id = DeploymentID(name="GangApp", app_name="gang_app")
+    controller = serve.context._get_global_client()._controller
     config_files = {
         replicas: os.path.join(
             os.path.dirname(__file__),
@@ -845,48 +848,32 @@ def test_deploy_gang_scaling(serve_instance, initial_replicas, final_replicas):
         for replicas in (initial_replicas, final_replicas)
     }
 
-    def wait_for_all_replicas_running(num_replicas):
-        # Wait for every replica to be RUNNING (not just the app status);
-        # otherwise requests can route to a subset of ready replicas and miss
-        # other gangs entirely, which causes flakiness.
-        def verify_running():
-            status = serve.status()
-            app_status = status.applications.get("gang_app")
-            if app_status is None or app_status.status != "RUNNING":
-                return False
-            dep = list(app_status.deployments.values())[0]
-            return dep.replica_states.get("RUNNING", 0) == num_replicas
-
-        wait_for_condition(verify_running, timeout=30)
-
     def collect_gang_ids(num_replicas):
-        # Poll until every replica has served at least one request, so we are
-        # guaranteed to observe every gang regardless of routing skew.
-        app_url = get_application_url(app_name="gang_app")
-        gang_ids = set()
-        seen_pids = set()
-        for _ in range(num_replicas * 50):
-            data = json.loads(httpx.post(f"{app_url}/").text)
-            if data["gang_id"] is not None:
-                gang_ids.add(data["gang_id"])
-            seen_pids.add(data["pid"])
-            if len(seen_pids) >= num_replicas:
-                break
-        assert (
-            len(seen_pids) == num_replicas
-        ), f"Only reached {len(seen_pids)}/{num_replicas} replicas"
-        return gang_ids
+        # Query the controller directly for replica gang contexts. Probing
+        # via HTTP is unreliable for observing every replica: under direct
+        # ingress, get_application_url returns a single replica's port, so
+        # all requests land on one replica.
+        collected = []
+
+        def _ready():
+            replicas = ray.get(
+                controller._dump_replica_states_for_testing.remote(deployment_id)
+            )
+            running = replicas.get([ReplicaState.RUNNING])
+            if len(running) != num_replicas:
+                return False
+            collected.append({r.gang_context.gang_id for r in running})
+            return True
+
+        wait_for_condition(_ready, timeout=30)
+        return collected[-1]
 
     subprocess.check_output(["serve", "deploy", config_files[initial_replicas]])
-    wait_for_all_replicas_running(initial_replicas)
-
     initial_num_gangs = initial_replicas // gang_size
     initial_gang_ids = collect_gang_ids(initial_replicas)
     assert len(initial_gang_ids) == initial_num_gangs
 
     subprocess.check_output(["serve", "deploy", config_files[final_replicas]])
-    wait_for_all_replicas_running(final_replicas)
-
     final_num_gangs = final_replicas // gang_size
     final_gang_ids = collect_gang_ids(final_replicas)
     assert len(final_gang_ids) == final_num_gangs

--- a/python/ray/serve/tests/test_cli.py
+++ b/python/ray/serve/tests/test_cli.py
@@ -845,41 +845,50 @@ def test_deploy_gang_scaling(serve_instance, initial_replicas, final_replicas):
         for replicas in (initial_replicas, final_replicas)
     }
 
-    # First deploy.
-    subprocess.check_output(["serve", "deploy", config_files[initial_replicas]])
-    app_url = get_application_url(app_name="gang_app")
-    wait_for_condition(lambda: httpx.post(f"{app_url}/").status_code == 200, timeout=30)
+    def wait_for_all_replicas_running(num_replicas):
+        # Wait for every replica to be RUNNING (not just the app status);
+        # otherwise requests can route to a subset of ready replicas and miss
+        # other gangs entirely, which causes flakiness.
+        def verify_running():
+            status = serve.status()
+            app_status = status.applications.get("gang_app")
+            if app_status is None or app_status.status != "RUNNING":
+                return False
+            dep = list(app_status.deployments.values())[0]
+            return dep.replica_states.get("RUNNING", 0) == num_replicas
 
-    # Collect initial gang IDs.
+        wait_for_condition(verify_running, timeout=30)
+
+    def collect_gang_ids(num_replicas):
+        # Poll until every replica has served at least one request, so we are
+        # guaranteed to observe every gang regardless of routing skew.
+        app_url = get_application_url(app_name="gang_app")
+        gang_ids = set()
+        seen_pids = set()
+        for _ in range(num_replicas * 50):
+            data = json.loads(httpx.post(f"{app_url}/").text)
+            if data["gang_id"] is not None:
+                gang_ids.add(data["gang_id"])
+            seen_pids.add(data["pid"])
+            if len(seen_pids) >= num_replicas:
+                break
+        assert (
+            len(seen_pids) == num_replicas
+        ), f"Only reached {len(seen_pids)}/{num_replicas} replicas"
+        return gang_ids
+
+    subprocess.check_output(["serve", "deploy", config_files[initial_replicas]])
+    wait_for_all_replicas_running(initial_replicas)
+
     initial_num_gangs = initial_replicas // gang_size
-    initial_gang_ids = set()
-    for _ in range(initial_replicas * 10):
-        data = json.loads(httpx.post(f"{app_url}/").text)
-        if data["gang_id"] is not None:
-            initial_gang_ids.add(data["gang_id"])
+    initial_gang_ids = collect_gang_ids(initial_replicas)
     assert len(initial_gang_ids) == initial_num_gangs
 
-    # Redeploy with new replica count.
     subprocess.check_output(["serve", "deploy", config_files[final_replicas]])
+    wait_for_all_replicas_running(final_replicas)
 
-    # Wait for rescale to complete.
-    def rescale_complete():
-        status = serve.status()
-        app_status = status.applications.get("gang_app")
-        if app_status is None or app_status.status != "RUNNING":
-            return False
-        dep = list(app_status.deployments.values())[0]
-        return dep.replica_states.get("RUNNING", 0) == final_replicas
-
-    wait_for_condition(rescale_complete, timeout=30)
-
-    # Collect final gang IDs.
     final_num_gangs = final_replicas // gang_size
-    final_gang_ids = set()
-    for _ in range(final_replicas * 10):
-        data = json.loads(httpx.post(f"{app_url}/").text)
-        if data["gang_id"] is not None:
-            final_gang_ids.add(data["gang_id"])
+    final_gang_ids = collect_gang_ids(final_replicas)
     assert len(final_gang_ids) == final_num_gangs
 
     # The smaller set of gangs must be a subset of the larger set,

